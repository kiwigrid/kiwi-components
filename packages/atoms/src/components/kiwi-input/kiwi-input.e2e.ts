import { newE2EPage } from '@stencil/core/testing';

describe('kiwi-input', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<kiwi-input></kiwi-input>');

    const element = await page.find('kiwi-input');
    expect(element).toHaveClass('hydrated');
  });
});
