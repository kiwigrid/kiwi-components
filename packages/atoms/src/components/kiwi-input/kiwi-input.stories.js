import { html } from 'lit-html';
import notes from './readme.md';

export default {
  title: 'KiwiComponents/Input',
  parameters: { docs: { description: { component: notes } } },
  decorators: [],
};

export const basic = (args) => {
  const { name, placeholder, value } = args;

  return html`<div class="panel panel-default">
    <div class="panel-body">
      <kiwi-input
        name=${name}
        placeholder=${placeholder}
        value=${value}
      ></kiwi-input>
    </div>
  </div>`;
};
basic.args = {
  name: 'inputName',
  placeholder: 'Type something here...',
  value: '',
};
