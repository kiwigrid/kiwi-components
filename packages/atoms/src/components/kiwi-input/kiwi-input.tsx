import { Component, ComponentInterface, h, JSX, Prop } from '@stencil/core';

@Component({
  tag: 'kiwi-input',
  shadow: false,
})
export class KiwiInput implements ComponentInterface {
  /**
   * The _name_ attribute of the underlying input.
   */
  @Prop()
  name?: string;

  /**
   * The _placeholder_ attribute of the underlying input.
   * Default _empty_.
   */
  @Prop()
  placeholder = '';

  /**
   * The _value_ attribute of the underlying input.
   * Default _empty_.
   */
  @Prop()
  value = '';

  private input!: HTMLInputElement;

  render(): JSX.Element {
    return (
      <input
        ref={(el) => (this.input = el as HTMLInputElement)}
        type="text"
        name={this.name}
        placeholder={this.placeholder}
        value={this.value}
        onInput={this.handleOnInput}
        class="form-control"
      />
    );
  }

  private handleOnInput = (): void => {
    this.value = this.input.value;
  };
}
