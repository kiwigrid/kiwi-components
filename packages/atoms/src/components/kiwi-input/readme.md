# kiwi-input



<!-- Auto Generated Below -->


## Properties

| Property      | Attribute     | Description                                                           | Type                  | Default     |
| ------------- | ------------- | --------------------------------------------------------------------- | --------------------- | ----------- |
| `name`        | `name`        | The _name_ attribute of the underlying input.                         | `string \| undefined` | `undefined` |
| `placeholder` | `placeholder` | The _placeholder_ attribute of the underlying input. Default _empty_. | `string`              | `''`        |
| `value`       | `value`       | The _value_ attribute of the underlying input. Default _empty_.       | `string`              | `''`        |


## Dependencies

### Used by

 - [kiwi-searchbar](../kiwi-searchbar)

### Graph
```mermaid
graph TD;
  kiwi-searchbar --> kiwi-input
  style kiwi-input fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
