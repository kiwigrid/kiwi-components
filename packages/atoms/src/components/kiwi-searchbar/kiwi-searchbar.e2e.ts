import { newE2EPage } from '@stencil/core/testing';

describe('kiwi-searchbar', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<kiwi-searchbar></kiwi-searchbar>');

    const element = await page.find('kiwi-searchbar');
    expect(element).toHaveClass('hydrated');
  });
});
