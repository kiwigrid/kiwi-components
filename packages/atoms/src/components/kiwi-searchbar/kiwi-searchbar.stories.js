import { html } from 'lit-html';
import notes from './readme.md';

export default {
  title: 'KiwiComponents/Searchbar',
  parameters: { docs: { description: { component: notes } } },
  decorators: [],
};

export const basic = (args) => {
  const { debounce, placeholder } = args;

  return html` <div class="panel panel-default">
    <div class="panel-body">
      <kiwi-searchbar debounce=${debounce} placeholder=${placeholder} />
    </div>
  </div>`;
};
basic.args = {
  debounce: 400,
  placeholder: 'Search...',
};
