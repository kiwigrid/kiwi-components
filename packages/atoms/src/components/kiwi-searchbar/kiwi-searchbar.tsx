import {
  Component,
  ComponentInterface,
  Event,
  EventEmitter,
  h,
  JSX,
  Prop,
} from '@stencil/core';
import { Subject } from 'rxjs';
import { debounceTime, tap } from 'rxjs/operators';

@Component({
  tag: 'kiwi-searchbar',
  shadow: false,
})
export class KiwiSearchbar implements ComponentInterface {
  /**
   * The debounce time for the `triggerSearch` event in milliseconds.
   * Default: `400`
   */
  @Prop()
  debounce = 400;

  /**
   * The placeholder to use for the search input.
   */
  @Prop()
  placeholder = '';

  /** The input value */
  @Prop()
  value = '';

  /**
   * Event containing the search query.
   * Fired after `debounce` time, either on type, or on click (immediately).
   */
  @Event() triggerSearch!: EventEmitter<string>;

  private kiwiInput!: HTMLKiwiInputElement;
  private debouncedEmission: Subject<string> = new Subject();

  componentWillLoad(): void {
    this.debouncedEmission
      .pipe(
        tap((query) => {
          console.log('Query vor debounce:', query, this.debounce);
        }),
        debounceTime(this.debounce),
        tap((query) => {
          console.log('Query nach debounce:', query);
        }),
      )
      .subscribe((query) => this.emitTriggerSearch(query));
  }

  render(): JSX.Element {
    return (
      <div class="input-group">
        <kiwi-input
          ref={(el) => (this.kiwiInput = el as HTMLKiwiInputElement)}
          onInput={this.handleInput}
          name="search"
          placeholder={this.placeholder}
          value={this.value}
        ></kiwi-input>
        <span class="input-group-btn">
          {this.value && (
            <button
              aria-label="Clear"
              onClick={this.handleRemoveClick}
              class="btn btn-default btn-addon"
              type="button"
            >
              <span class="glyphicon glyphicon-remove"></span>
            </button>
          )}
          <button
            aria-label="Search"
            onClick={this.handleSearchClick}
            class="btn btn-primary btn-addon"
            type="button"
          >
            <span class="glyphicon glyphicon-search"></span>
          </button>
        </span>
      </div>
    );
  }

  private handleInput: () => void = () =>
    this.debouncedEmission.next(this.kiwiInput.value);

  private handleSearchClick: () => void = () =>
    this.emitTriggerSearch(this.kiwiInput.value);

  private handleRemoveClick: () => void = () => this.emitTriggerSearch('');

  private emitTriggerSearch(query: string): void {
    this.triggerSearch.emit(query);
  }
}
