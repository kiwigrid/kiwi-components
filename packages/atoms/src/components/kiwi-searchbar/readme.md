# kiwi-searchbar



<!-- Auto Generated Below -->


## Properties

| Property      | Attribute     | Description                                                                     | Type     | Default |
| ------------- | ------------- | ------------------------------------------------------------------------------- | -------- | ------- |
| `debounce`    | `debounce`    | The debounce time for the `triggerSearch` event in milliseconds. Default: `400` | `number` | `400`   |
| `placeholder` | `placeholder` | The placeholder to use for the search input.                                    | `string` | `''`    |
| `value`       | `value`       | The input value                                                                 | `string` | `''`    |


## Events

| Event           | Description                                                                                                | Type                  |
| --------------- | ---------------------------------------------------------------------------------------------------------- | --------------------- |
| `triggerSearch` | Event containing the search query. Fired after `debounce` time, either on type, or on click (immediately). | `CustomEvent<string>` |


## Dependencies

### Depends on

- [kiwi-input](../kiwi-input)

### Graph
```mermaid
graph TD;
  kiwi-searchbar --> kiwi-input
  style kiwi-searchbar fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
