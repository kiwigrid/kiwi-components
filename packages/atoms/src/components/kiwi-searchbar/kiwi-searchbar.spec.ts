import { newSpecPage } from '@stencil/core/testing';
import { fakeSchedulers } from 'rxjs-marbles/jest';
import { KiwiInput } from '../kiwi-input/kiwi-input';
import { KiwiSearchbar } from './kiwi-searchbar';

describe('kiwi-searchbar', () => {
  it('builds', () => {
    const searchbar = new KiwiSearchbar();
    expect(searchbar).toBeTruthy();
    expect(searchbar.debounce).toEqual(400);
  });

  it('renders', async () => {
    const searchbar = await newSpecPage({
      components: [KiwiSearchbar, KiwiInput],
      html: `<kiwi-searchbar></kiwi-searchbar>`,
    });

    expect(searchbar.root).toMatchSnapshot();
  });

  it('emits search event on button click', async () => {
    const searchbar = await newSpecPage({
      components: [KiwiSearchbar, KiwiInput],
      html: `<kiwi-searchbar></kiwi-searchbar>`,
    });

    const searchbarInstance = searchbar.rootInstance;

    jest.spyOn(searchbarInstance.triggerSearch, 'emit');

    searchbar.root!.querySelector('input')!.value = 'abc';
    searchbar
      .root!.querySelector<HTMLButtonElement>('button[aria-label="Search"]')!
      .click();

    expect(searchbarInstance.triggerSearch.emit).toHaveBeenCalled();
  });

  it('shows remove button', async () => {
    const searchbar = await newSpecPage({
      components: [KiwiSearchbar],
      html: `<kiwi-searchbar value="query"></kiwi-searchbar>`,
    });

    const searchbarInstance = searchbar.rootInstance;

    jest.spyOn(searchbarInstance.triggerSearch, 'emit');

    searchbar
      .root!.querySelector<HTMLButtonElement>('button[aria-label="Clear"]')!
      .click();

    expect(searchbarInstance.triggerSearch.emit).toBeCalledWith('');
  });

  describe('async behaviour', () => {
    beforeEach(() => jest.useFakeTimers());

    it(
      'emits search event on type',
      fakeSchedulers(async (advance) => {
        const searchbar = await newSpecPage({
          components: [KiwiSearchbar, KiwiInput],
          html: `<kiwi-searchbar debounce="1000"></kiwi-searchbar>`,
        });

        const triggerSearchSpy = jest.fn();
        searchbar.root!.addEventListener('triggerSearch', triggerSearchSpy);

        const input = searchbar.root!.querySelector('kiwi-input');
        input!.value = '123';
        input!.dispatchEvent(new Event('input'));
        await searchbar.waitForChanges();

        expect(triggerSearchSpy).not.toHaveBeenCalled();
        advance(800);
        expect(triggerSearchSpy).not.toHaveBeenCalled();
        advance(200);
        expect(triggerSearchSpy).toHaveBeenCalledWith(
          expect.objectContaining({ detail: '123' }),
        );
      }),
    );
  });
});
