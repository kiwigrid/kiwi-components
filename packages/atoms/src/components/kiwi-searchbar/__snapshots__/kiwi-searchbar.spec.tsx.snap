// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kiwi-searchbar renders 1`] = `
<kiwi-searchbar>
  <div class="input-group">
    <kiwi-input>
      <input class="form-control" name="search" placeholder="" type="text" value="">
    </kiwi-input>
    <span class="input-group-btn">
      <button aria-label="Search" class="btn btn-addon btn-primary" type="button">
        <span class="glyphicon glyphicon-search"></span>
      </button>
    </span>
  </div>
</kiwi-searchbar>
`;
